storage-s3-dynamodb/src/main/java/io/delta/storage/RetryableCloseableIterator.java

```diff
@@ -0,0 +1,203 @@
+import java.io.UncheckedIOException;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import org.apache.hadoop.fs.s3a.RemoteFileChangedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class presents an iterator view over the iterator supplier in the constructor.
+ *
+ * This class assumes that the iterator supplied by the supplier can throw, and that subsequent
+ * supplier.get() calls will return an iterator over the same data.
+ *
+ * If there are any RemoteFileChangedException during `next` and `hasNext` calls, will retry
+ * at most `MAX_RETRIES` times. If there are similar exceptions during the retry, those are handled
+ * and count towards the MAX_RETRIES.
+ *
+ * Internally, keeps track of the last-successfully-returned index. Upon retry, will iterate back
+ * to that same position.
+ */
+public class RetryableCloseableIterator implements CloseableIterator<String> {
+    private static final Logger LOG = LoggerFactory.getLogger(RetryableCloseableIterator.class);
+
+    public static final int DEFAULT_MAX_RETRIES = 3;
+
+    private final Supplier<CloseableIterator<String>> iterSupplier;
+
+    private final int maxRetries;
+
+    /**
+     * Index of the last element successfully returned without an exception. A value of -1 means
+     * that no element has ever been returned yet.
+     */
+    private int lastSuccessfullIndex;
+
+    private int numRetries = 0;
+
+    private CloseableIterator<String> currentIter;
+
+    public RetryableCloseableIterator(
+            Supplier<CloseableIterator<String>> iterSupplier,
+            int maxRetries) {
+        if (maxRetries < 0) throw new IllegalArgumentException("maxRetries can't be negative");
+
+        this.iterSupplier = Objects.requireNonNull(iterSupplier);
+        this.maxRetries = maxRetries;
+        this.lastSuccessfullIndex = -1;
+        this.currentIter = this.iterSupplier.get();
+    }
+
+    public RetryableCloseableIterator(Supplier<CloseableIterator<String>> iterSupplier) {
+        this(iterSupplier, DEFAULT_MAX_RETRIES);
+    }
+
+    /** Visible for testing. */
+    public int getLastSuccessfullIndex() {
+        return lastSuccessfullIndex;
+    }
+
+    /** Visible for testing. */
+    public int getNumRetries() {
+        return numRetries;
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (currentIter != null) {
+            currentIter.close();
+        }
+    }
+
+    /**
+     * `hasNext` must be idempotent. It does not change the `lastSuccessfulIndex` variable.
+     */
+    @Override
+    public boolean hasNext() {
+        try {
+            return hasNextInternal();
+        } catch (IOException ex) {
+            // `endsWith` should still work if the class is shaded.
+            final String exClassName = ex.getClass().getName();
+            if (exClassName.endsWith("org.apache.hadoop.fs.s3a.RemoteFileChangedException")) {
```

Comment:
Could you create a method to avoid code duplication?
```
private boolean isRemoteFileChangedException(IOException ex) {
  // `endsWith` should still work if the class is shaded.
  final String exClassName = ex.getClass().getName();
  return exClassName.endsWith("org.apache.hadoop.fs.s3a.RemoteFileChangedException");
}

```